### PR TITLE
feat(sql): implement level 2 order book price function

### PR DESCRIPTION
**Pull request by nwoolmer**
Closes https://github.com/questdb/questdb/issues/4454

Implements the `L2PRICE(targetSize, size_0, value_0, size_1, value_1, size_n, value_n)` function.

This calculates an average price for fulfilling `targetSize` based on the order book sizes and values provided.

image

Docs here: https://questdb.io/docs/reference/function/finance/

### CODE REVIEW

File: core/src/main/java/io/questdb/griffin/engine/functions/finance/LevelTwoPriceFunctionFactory.java [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+    public String getSignature() {
+        return "l2price(LV)";
+    }
+
+    @Override
+    public Function newInstance(int position, ObjList<Function> args, IntList argPositions, CairoConfiguration configuration, SqlExecutionContext sqlExecutionContext) throws SqlException {
+        if (args.size() % 2 == 0) {
+            throw SqlException.position(argPositions.getLast()).put("l2price requires an odd number of arguments.");
+        }
+
```
Comment (ideoma) [inline core/src/main/java/io/questdb/griffin/engine/functions/finance/LevelTwoPriceFunctionFactory.java:52]: Please check the parameter types.
Everything should be double or a type that can be implicitly cast to double.

File: core/src/main/java/io/questdb/griffin/engine/functions/finance/LevelTwoPriceFunctionFactory.java [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+        final int numberOfPairs = (args.size() - 1) / 2;
+        final Function target = args.getQuick(0);
+
+        if (target.isNullConstant()) {
+            throw SqlException.position(argPositions.getQuick(0)).put("l2price requires a non-null first argument.");
```
Comment (ideoma) [inline core/src/main/java/io/questdb/griffin/engine/functions/finance/LevelTwoPriceFunctionFactory.java:57]: Not an error imho, we can shortcut here to return NULL

File: core/src/main/java/io/questdb/griffin/engine/functions/finance/LevelTwoPriceFunctionFactory.java [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+                ta += size;
+
+                // if order not fulfilled
+                if (ta < t) {
+                    pa += (size * value);
+                    rt -= size;
+                    continue;
+                }
+
+                if (ta >= t) {
```
Comment (ideoma) [inline core/src/main/java/io/questdb/griffin/engine/functions/finance/LevelTwoPriceFunctionFactory.java:322]: This `if` is not necessary as far as I can see. I'd rewrite the `if`s like this instead

```java
if (ta >= t) {
                   if (i == 1) {
                        return value;
                    }

                    pp = rt * value;
                    return (pa + pp) / t;
} else {
                    pa += (size * value);
                    rt -= size;
}

File: core/src/main/java/io/questdb/griffin/engine/functions/finance/LevelTwoPriceFunctionFactory.java [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+import io.questdb.std.IntList;
+import io.questdb.std.ObjList;
+
+/**
+ * Expects a target size, and then pairs of (order size, order amount).
+ */
+public class LevelTwoPriceFunctionFactory implements FunctionFactory {
+    @Override
+    public String getSignature() {
+        return "l2price(LV)";
```
Comment (ideoma) [inline core/src/main/java/io/questdb/griffin/engine/functions/finance/LevelTwoPriceFunctionFactory.java:44]:
```suggestion
        return "l2price(DV)";
```

File: core/src/test/java/io/questdb/test/griffin/engine/functions/finance/LevelTwoPriceFunctionFactoryTest.java [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+ ******************************************************************************/
+
+package io.questdb.test.griffin.engine.functions.finance;
+
+import io.questdb.griffin.FunctionFactory;
+import io.questdb.griffin.engine.functions.finance.LevelTwoPriceFunctionFactory;
+import io.questdb.test.griffin.engine.AbstractFunctionFactoryTest;
+import org.junit.Test;
+
+public class LevelTwoPriceFunctionFactoryTest extends AbstractFunctionFactoryTest {
```
Comment (ideoma) [inline core/src/test/java/io/questdb/test/griffin/engine/functions/finance/LevelTwoPriceFunctionFactoryTest.java:32]: Please add a few unhappy path tests, wrong types, and wrong param count.

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 97 / 119 (81.51%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/finance/LevelTwoPriceFunctionFactory.java|97|119|81.51%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 129 / 129 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/finance/LevelTwoPriceFunctionFactory.java|129|129|100.00%|

**Comment by javier** [comment]
Found this.

Create a table like

```
CREATE TABLE 'btc_trades' (
  symbol SYMBOL capacity 256 CACHE,
  side SYMBOL capacity 256 CACHE,
  price DOUBLE,
  amount DOUBLE,
  timestamp TIMESTAMP
) timestamp (timestamp) PARTITION BY DAY WAL DEDUP UPSERT KEYS(symbol, timestamp);
```

Now insert some data, as in:

```
INSERT INTO btc_trades (symbol, side, price, amount, timestamp) VALUES
('BTC-USD', 'buy', 25738.01, 0.00192304, '2023-09-05T16:59:45.804537Z'),
('BTC-USD', 'buy', 25738.01, 0.00254854, '2023-09-05T16:59:46.639102Z'),
('BTC-USD', 'buy', 25738.01, 0.00038735, '2023-09-05T16:59:47.229826Z'),
('BTC-USD', 'buy', 25738.01, 0.00739719, '2023-09-05T16:59:48.289720Z'),
('BTC-USD', 'buy', 25738.01, 0.00052228, '2023-09-05T16:59:48.829058Z'),
('BTC-USD', 'buy', 25738.01, 0.00071191, '2023-09-05T16:59:49.002844Z'),
('BTC-USD', 'buy', 25738.76, 0.00367218, '2023-09-05T16:59:51.153025Z'),
('BTC-USD', 'buy', 25741.68, 0.05034042, '2023-09-05T16:59:52.260796Z'),
('BTC-USD', 'buy', 25740.38, 0.00219286, '2023-09-05T16:59:53.713701Z'),
('BTC-USD', 'buy', 25739.75, 0.00037233, '2023-09-05T16:59:56.452825Z'),
('BTC-USD', 'buy', 25735.97, 0.00006303, '2023-09-05T16:59:58.095714Z'),
('BTC-USD', 'buy', 25735.97, 0.00816829, '2023-09-05T16:59:58.761737Z');

```

Now execute this query

```

with recent_trades as
(
    select *, 
    FIRST_VALUE(amount) over(partition by symbol order by timestamp rows between 1 preceding and 1 preceding) as amount1,
    FIRST_VALUE(amount) over(partition by symbol order by timestamp rows between 2 preceding and 2 preceding) as amount2,
    FIRST_VALUE(amount) over(partition by symbol order by timestamp rows between 3 preceding and 3 preceding) as amount3,
    FIRST_VALUE(amount) over(partition by symbol order by timestamp rows between 4 preceding and 4 preceding) as amount4,
    FIRST_VALUE(amount) over(partition by symbol order by timestamp rows between 5 preceding and 5 preceding) as amount5,
    FIRST_VALUE(amount) over(partition by symbol order by timestamp rows between 6 preceding and 6 preceding) as amount6,
    FIRST_VALUE(price) over(partition by symbol order by timestamp rows between 1 preceding and 1 preceding) as price1,
    FIRST_VALUE(price) over(partition by symbol order by timestamp rows between 2 preceding and 2 preceding) as price2,
    FIRST_VALUE(price) over(partition by symbol order by timestamp rows between 3 preceding and 3 preceding) as price3,
    FIRST_VALUE(price) over(partition by symbol order by timestamp rows between 4 preceding and 4 preceding) as price4,
    FIRST_VALUE(price) over(partition by symbol order by timestamp rows between 5 preceding and 5 preceding) as price5,
     FIRST_VALUE(price) over(partition by symbol order by timestamp rows between 6 preceding and 6 preceding) as price6
    from btc_trades
    where side = 'buy'
    limit -12
)
select *, l2price(0.0015, amount1, price1, amount2, price2, amount3, price3, 
amount4, price4, amount5, price5) from recent_trades
limit -1;
```

It all works. But if I add an extra amount and price, questdb crashes
```

with recent_trades as
(
    select *, 
    FIRST_VALUE(amount) over(partition by symbol order by timestamp rows between 1 preceding and 1 preceding) as amount1,
    FIRST_VALUE(amount) over(partition by symbol order by timestamp rows between 2 preceding and 2 preceding) as amount2,
    FIRST_VALUE(amount) over(partition by symbol order by timestamp rows between 3 preceding and 3 preceding) as amount3,
    FIRST_VALUE(amount) over(partition by symbol order by timestamp rows between 4 preceding and 4 preceding) as amount4,
    FIRST_VALUE(amount) over(partition by symbol order by timestamp rows between 5 preceding and 5 preceding) as amount5,
    FIRST_VALUE(amount) over(partition by symbol order by timestamp rows between 6 preceding and 6 preceding) as amount6,
    FIRST_VALUE(price) over(partition by symbol order by timestamp rows between 1 preceding and 1 preceding) as price1,
    FIRST_VALUE(price) over(partition by symbol order by timestamp rows between 2 preceding and 2 preceding) as price2,
    FIRST_VALUE(price) over(partition by symbol order by timestamp rows between 3 preceding and 3 preceding) as price3,
    FIRST_VALUE(price) over(partition by symbol order by timestamp rows between 4 preceding and 4 preceding) as price4,
    FIRST_VALUE(price) over(partition by symbol order by timestamp rows between 5 preceding and 5 preceding) as price5,
     FIRST_VALUE(price) over(partition by symbol order by timestamp rows between 6 preceding and 6 preceding) as price6
    from btc_trades
    where side = 'buy'
    limit -12
)
select *, l2price(0.0015, amount1, price1, amount2, price2, amount3, price3, 
amount4, price4, amount5, price5, amount6, price6) from recent_trades
limit -1;
```

I get the error in the log
```
2024-05-21T15:15:22.822028Z I i.q.c.h.p.JsonQueryProcessorState [171] execute-new [skip: 0, stop: 1000]
2024-05-21T15:15:22.823140Z C i.q.c.h.p.JsonQueryProcessorState [171] internal error [ex=
java.lang.UnsupportedOperationException
	at io.questdb.griffin.engine.functions.DoubleFunction.getLong(DoubleFunction.java:107)
	at io.questdb.griffin.engine.functions.finance.LevelTwoPriceFunctionFactory$L2PriceFunctionN.getDouble(LevelTwoPriceFunctionFactory.java:335)
	at io.questdb.cairo.sql.VirtualRecord.getDouble(VirtualRecord.java:93)
	at io.questdb.cutlass.http.processors.JsonQueryProcessorState.putDoubleValue(JsonQueryProcessorState.java:809)
	at io.questdb.cutlass.http.processors.JsonQueryProcessorState.doQueryRecord(JsonQueryProcessorState.java:562)
	at io.questdb.cutlass.http.processors.JsonQueryProcessorState.doRecordFetchLoop(JsonQueryProcessorState.java:697)
	at io.questdb.cutlass.http.processors.JsonQueryProcessorState.onSendRecordsLoop(JsonQueryProcessorState.java:780)
	at io.questdb.cutlass.http.processors.JsonQueryProcessorState.onQueryPrefix(JsonQueryProcessorState.java:751)
	at io.questdb.cutlass.http.processors.JsonQueryProcessorState.onSetupFirstRecord(JsonQueryProcessorState.java:800)
	at io.questdb.cutlass.http.processors.JsonQueryProcessorState.resume(JsonQueryProcessorState.java:945)
	at io.questdb.cutlass.http.processors.JsonQueryProcessor.doResumeSend(JsonQueryProcessor.java:344)
	at io.questdb.cutlass.http.processors.JsonQueryProcessor.executeSelect(JsonQueryProcessor.java:614)
	at io.questdb.cutlass.http.processors.JsonQueryProcessor.executeNewSelect(JsonQueryProcessor.java:578)
	at io.questdb.cutlass.http.processors.JsonQueryProcessor.compileAndExecuteQuery(JsonQueryProcessor.java:486)
	at io.questdb.cutlass.http.processors.JsonQueryProcessor.execute0(JsonQueryProcessor.java:199)
	at io.questdb.cutlass.http.processors.JsonQueryProcessor.onRequestComplete(JsonQueryProcessor.java:274)
	at io.questdb.cutlass.http.HttpConnectionContext.handleClientRecv(HttpConnectionContext.java:919)
	at io.questdb.cutlass.http.HttpConnectionContext.handleClientOperation(HttpConnectionContext.java:260)
	at io.questdb.cutlass.http.HttpServer.handleClientOperation(HttpServer.java:276)
	at io.questdb.cutlass.http.HttpServer$1.lambda$$0(HttpServer.java:88)
	at io.questdb.network.AbstractIODispatcher.processIOQueue(AbstractIODispatcher.java:199)
	at io.questdb.cutlass.http.HttpServer$1.run(HttpServer.java:92)
	at io.questdb.mp.Worker.run(Worker.java:151)
, q=`with recent_trades as
```


However, when I change the target price (first param in the function) to a value of 1 or higher, it works. It doesn't return any data for that function, as all the trades are priced under 1, but it does not break. With the decimal price it works fine but only until a number of parameters.

**Pull request merged**